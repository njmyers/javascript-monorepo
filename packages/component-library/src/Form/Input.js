// @flow
import * as React from 'react';
import BEM from '../BEM';

type Props = {
  type: string,
  name: string,
  label?: string,
  required: boolean,
  block: string,
  modifiers?: string | Array<string> | void,
  className?: string,
  focus: boolean,
  onChange?: Function,
  placeholder?: string,
  value: string | number,
};

class Input extends React.PureComponent<Props> {
  static defaultProps = {
    type: 'text',
    name: 'input',
    required: false,
    block: 'formInput',
    focus: false,
    placeholder: '',
  };

  // ref to this input
  input: ?HTMLInputElement;
  // label defaults to property name
  label = this.props.label ? this.props.label : this.props.name;

  componentDidMount() {
    // $FlowFixMe
    if (this.props.focus) this.input.focus();
  }

  render() {
    return (
      <BEM block={this.props.block}>
        <div modifiers={this.props.modifiers} className={this.props.className}>
          <label
            element="label"
            htmlFor={this.props.name}
            modifiers={this.props.modifiers}
          >
            {this.label}
          </label>
          <input
            ref={(node) => (this.input = node)}
            element="input"
            modifiers={this.props.modifiers}
            type={this.props.type}
            value={this.props.value}
            name={this.props.name}
            onChange={this.props.onChange}
            required={this.props.required}
            placeholder={this.props.placeholder}
          />
        </div>
      </BEM>
    );
  }
}

export default Input;
