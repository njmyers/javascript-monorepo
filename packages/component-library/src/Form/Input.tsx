// @ts-nocheck
import * as React from 'react';
import Element from './Element';
import { Props } from './Types';

const Input = React.forwardRef((props: Props, ref) => (
  <Element as="input" forwardedRef={ref} {...props} />
));

Input.defaultProps = {
  type: 'text',
  name: 'input',
  required: false,
  block: 'formInput',
  placeholder: '',
};

export default Input;
