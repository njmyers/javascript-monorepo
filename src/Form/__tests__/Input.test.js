import React from 'react';
import Input from '../Input';

import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('the input works', () => {
  test('the input mounts without crashing', () => {
    const wrapper = shallow(<Input />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('the input receives value as prop', () => {
    const wrapper = shallow(<Input value={4} />);
    expect(wrapper.find('input').props().value).toBe(4);
  });

  test('the input receives value changes as prop', () => {
    const wrapper = shallow(<Input value={4} />);
    expect(wrapper.find('input').props().value).toBe(4);

    wrapper.setProps({ value: 5 });
    expect(wrapper.find('input').props().value).toBe(5);
  });

  test('the input passes name, type and label property', () => {
    const wrapper = shallow(<Input name="age" type="number" label="Age" />);
    expect(wrapper.find('input').props().name).toBe('age');
    expect(wrapper.find('input').props().type).toBe('number');
    expect(wrapper.find('label').text()).toBe('Age');
  });

  test('the input passes overwrites label with name when no prop is provided', () => {
    const wrapper = shallow(<Input name="age" type="number" />);
    expect(wrapper.find('input').props().name).toBe('age');
    expect(wrapper.find('input').props().type).toBe('number');
    expect(wrapper.find('label').text()).toBe('age');
  });

  test('the input renders default classNames', () => {
    const wrapper = mount(<Input />);
    expect(wrapper.find('div').props().className).toBe('formInput');
    expect(wrapper.find('label').props().className).toBe('formInput_label');
    expect(wrapper.find('input').props().className).toBe('formInput_input');
  });

  test('the input renders input classNames', () => {
    const wrapper = mount(<Input block="addressInput" />);
    expect(wrapper.find('div').props().className).toBe('addressInput');
    expect(wrapper.find('label').props().className).toBe('addressInput_label');
    expect(wrapper.find('input').props().className).toBe('addressInput_input');
  });

  test('the input renders adds modifier classNames', () => {
    const wrapper = mount(<Input block="addressInput" modifiers="active" />);
    expect(wrapper.find('div').props().className).toBe(
      'addressInput addressInput-active'
    );
    expect(wrapper.find('label').props().className).toBe(
      'addressInput_label addressInput_label-active'
    );
    expect(wrapper.find('input').props().className).toBe(
      'addressInput_input addressInput_input-active'
    );
  });

  test('the input renders adds multiple modifier classNames', () => {
    const wrapper = mount(
      <Input block="addressInput" modifiers={['active', 'red']} />
    );
    expect(wrapper.find('div').props().className).toBe(
      'addressInput addressInput-active addressInput-red'
    );
  });
});
