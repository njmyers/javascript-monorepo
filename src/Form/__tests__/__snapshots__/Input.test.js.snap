// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the input works the input mounts without crashing 1`] = `"<div class=\\"formInput\\"><label element=\\"label\\" for=\\"input\\" class=\\"formInput_label\\">input</label><input type=\\"text\\" element=\\"input\\" name=\\"input\\" placeholder=\\"\\" class=\\"formInput_input\\"/></div>"`;
